source/extensions/filters/http/oauth2/oauth_client.cc

```diff
@@ -108,8 +108,10 @@ void OAuth2ClientImpl::dispatchRequest(Http::RequestMessagePtr&& msg) {
   if (thread_local_cluster != nullptr) {
     in_flight_request_ = thread_local_cluster->httpAsyncClient().send(
         std::move(msg), *this,
-        Http::AsyncClient::RequestOptions().setTimeout(
-            std::chrono::milliseconds(PROTOBUF_GET_MS_REQUIRED(uri_, timeout))));
+        Http::AsyncClient::RequestOptions()
+            .setTimeout(std::chrono::milliseconds(PROTOBUF_GET_MS_REQUIRED(uri_, timeout)))
+            .setRetryPolicy(retry_policy_)
+            .setBufferBodyForRetry(true));
```

Comment:
Only set the retry policy and buffer flag when we actually have a valid retry policy.